Implement memory-mapped IO and multi-threading for BLAKE3 hashing

silvanshade opened the pull request:
This PR implements memory-mapped IO and multi-threading for BLAKE3 hashing.

Performance with these changes is now on par with the proposed Rust interop: https://github.com/NixOS/nix/pull/12416.

### Benchmarks

<details>
NOTE: These numbers are taken from https://github.com/NixOS/nix/pull/12416 based on the Rust implementation, and so are only an estimate. In practice they should (for the BLAKE3 results) be nearly identical, based on my local testing and original testing for the `libblake3` TBB feature upstream.


NOTE: The non-BLAKE3 results may be faster by a small margin with this PR than what is stated in the benchmarks since they will also make use of the memory-mapping changes.

### Config

CPU: AMD Ryzen 9 7950X 16-Core @ 5.88 GHz
RAM: 96GB @ 6400 MT/s
OS: CachyOS February 2025 release w/ bpfland scx

Input files created with:

```
head -c <size> /dev/urandom > ~/<size>.bin
```

example:

```
head -c 1G /dev/urandom > 1G.bin
```

Benchmarks all used the following:

```
hyperfine --warmup 3 './outputs/out/bin/nix hash file --type <algo> <file>'
```

### 100K file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/100K.bin
  Time (mean ± σ):       9.5 ms ±   0.1 ms    [User: 5.6 ms, System: 3.6 ms]
  Range (min … max):     9.1 ms …  10.0 ms    299 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/100K.bin
  Time (mean ± σ):       9.6 ms ±   0.9 ms    [User: 5.9 ms, System: 3.3 ms]
  Range (min … max):     9.2 ms …  24.7 ms    285 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/100K.bin
  Time (mean ± σ):       9.6 ms ±   0.9 ms    [User: 5.7 ms, System: 3.5 ms]
  Range (min … max):     9.0 ms …  23.1 ms    290 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/100K.bin
  Time (mean ± σ):       9.6 ms ±   0.9 ms    [User: 5.8 ms, System: 3.3 ms]
  Range (min … max):     9.0 ms …  23.3 ms    288 runs
```

### 10M file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/10M.bin
  Time (mean ± σ):      11.9 ms ±   2.6 ms    [User: 6.7 ms, System: 4.5 ms]
  Range (min … max):    11.0 ms …  32.8 ms    240 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/10M.bin
  Time (mean ± σ):      12.9 ms ±   1.1 ms    [User: 9.9 ms, System: 22.7 ms]
  Range (min … max):    11.3 ms …  16.6 ms    215 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/10M.bin
  Time (mean ± σ):      13.9 ms ±   0.4 ms    [User: 9.4 ms, System: 4.1 ms]
  Range (min … max):    13.3 ms …  16.3 ms    201 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/10M.bin
  Time (mean ± σ):      18.2 ms ±   0.5 ms    [User: 13.3 ms, System: 4.4 ms]
  Range (min … max):    17.4 ms …  21.6 ms    162 runs
```

### 100M file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/100M.bin
  Time (mean ± σ):      26.2 ms ±   0.8 ms    [User: 17.0 ms, System: 8.7 ms]
  Range (min … max):    24.9 ms …  29.1 ms    111 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/100M.bin
  Time (mean ± σ):      17.5 ms ±   1.6 ms    [User: 33.9 ms, System: 41.7 ms]
  Range (min … max):    15.2 ms …  23.8 ms    128 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/100M.bin
  Time (mean ± σ):      54.1 ms ±   0.5 ms    [User: 44.0 ms, System: 9.5 ms]
  Range (min … max):    53.4 ms …  55.5 ms    55 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/100M.bin
  Time (mean ± σ):      96.1 ms ±   0.9 ms    [User: 85.8 ms, System: 9.4 ms]
  Range (min … max):    95.1 ms …  98.5 ms    31 runs
```

### 300M file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/300M.bin
  Time (mean ± σ):      59.2 ms ±   0.9 ms    [User: 37.7 ms, System: 20.8 ms]
  Range (min … max):    57.8 ms …  61.9 ms    49 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/300M.bin
  Time (mean ± σ):      26.0 ms ±   1.6 ms    [User: 85.6 ms, System: 65.8 ms]
  Range (min … max):    22.8 ms …  31.1 ms    104 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/300M.bin
  Time (mean ± σ):     139.6 ms ±   0.8 ms    [User: 116.4 ms, System: 22.5 ms]
  Range (min … max):   138.1 ms … 141.0 ms    21 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/300M.bin
  Time (mean ± σ):     263.5 ms ±   3.0 ms    [User: 238.0 ms, System: 22.9 ms]
  Range (min … max):   260.4 ms … 269.5 ms    11 runs
```

### 1G file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/1G.bin
  Time (mean ± σ):     190.9 ms ±   1.5 ms    [User: 113.6 ms, System: 76.1 ms]
  Range (min … max):   188.8 ms … 194.3 ms    15 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/1G.bin
  Time (mean ± σ):      52.5 ms ±   5.0 ms    [User: 304.9 ms, System: 114.4 ms]
  Range (min … max):    50.0 ms …  88.9 ms    58 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/1G.bin
  Time (mean ± σ):     465.0 ms ±   4.5 ms    [User: 384.8 ms, System: 77.4 ms]
  Range (min … max):   461.8 ms … 477.0 ms    10 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/1G.bin
  Time (mean ± σ):     877.5 ms ±   8.9 ms    [User: 795.5 ms, System: 77.3 ms]
  Range (min … max):   870.8 ms … 900.8 ms    10 runs
```

### 20G file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/20G.bin
  Time (mean ± σ):      3.155 s ±  0.009 s    [User: 2.236 s, System: 0.914 s]
  Range (min … max):    3.143 s …  3.168 s    10 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/20G.bin
  Time (mean ± σ):     574.1 ms ±   9.8 ms    [User: 8339.7 ms, System: 1430.7 ms]
  Range (min … max):   563.6 ms … 596.4 ms    10 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/20G.bin
  Time (mean ± σ):      8.756 s ±  0.011 s    [User: 7.812 s, System: 0.933 s]
  Range (min … max):    8.737 s …  8.767 s    10 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/20G.bin
  Time (mean ± σ):     17.280 s ±  0.077 s    [User: 16.301 s, System: 0.954 s]
  Range (min … max):   17.220 s … 17.395 s    10 runs
```

### 64G file

#### BLAKE3 (original)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/64G.bin
  Time (mean ± σ):     17.145 s ±  0.086 s    [User: 7.154 s, System: 9.578 s]
  Range (min … max):   17.018 s … 17.276 s    10 runs
```

#### BLAKE3 (memory-mapping + tbb)
```
Benchmark 1: ./outputs/out/bin/nix hash file --type blake3 ~/64G.bin
  Time (mean ± σ):      1.822 s ±  0.011 s    [User: 26.895 s, System: 4.875 s]
  Range (min … max):    1.802 s …  1.832 s    10 runs
```

#### SHA256
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha256 ~/64G.bin
  Time (mean ± σ):     27.455 s ±  0.066 s    [User: 24.323 s, System: 3.072 s]
  Range (min … max):   27.343 s … 27.554 s    10 runs
```

#### SHA512
```
Benchmark 1: ./outputs/out/bin/nix hash file --type sha512 ~/64G.bin
  Time (mean ± σ):     53.807 s ±  0.212 s    [User: 50.615 s, System: 3.118 s]
  Range (min … max):   53.446 s … 54.187 s    10 runs
```
</details>

## Motivation

This PR adds additional functionality to the existing BLAKE3 implementation in `nix` to bring the performance on par with `b3sum`.

The performance difference between the two is due `b3sum` making use of the Rust BLAKE3 implementation which uses both memory-mapped IO and multi-threading.

Until recently, multi-threading was not available for the C-based `libblake3` but is now supported in [release `1.7.0`](https://github.com/BLAKE3-team/BLAKE3/releases/tag/1.7.0).

## Context

This PR is a follow up to https://github.com/NixOS/nix/pull/12379#issuecomment-2637977897.

Related: https://github.com/NixOS/nixpkgs/pull/390458

## Design Considerations

This PR implements memory-mapped IO via [`boost::iostreams::mapped_file`](https://www.boost.org/doc/libs/1_87_0/libs/iostreams/doc/classes/mapped_file.html), which adds a boost component dependency for `iostreams`.

Enabling multi-threading for `libblake3` also adds a dependency on `tbb` of at least version `2021_11`.

Memory-mapping is performed in:

```c++
void readFile(const Path & path, Sink & sink, bool memory_map = true);
```

and a new optional parameter `memory_map` is used to control whether memory-mapping is skipped in favor of normal file reading. (If memory-mapping fails, normal file reading is also used as the fallback).

This makes memory-mapping the default, which likely has performance implications beyond hashing. I would expect this to often be more performant than the alternative given available resources and modern hardware but haven't tested beyond hashing.

It may be appropriate to only enable memory-mapping when explicitly requested and/or gate memory-mapping behind an experimental feature. I can make those changes if requested.

@Ericson2314 @edolstra

Mic92 commented on packaging/dependencies.nix:
```diff
@@ -65,6 +66,18 @@ scope: {
       });
 
+  libblake3 = pkgs.libblake3.overrideAttrs (attrs: rec {
+    version = "1.7.0";
```
Can you also bump this in nixpkgs?

silvanshade replied:
@Mic92 Yes, there is a pending PR for that here: https://github.com/NixOS/nixpkgs/pull/390458

Ericson2314 replied:
I merged it

silvanshade commented on packaging/dependencies.nix:
```diff
@@ -65,6 +66,18 @@ scope: {
+      hash = "sha256-08D5hnU3I0VJ+RM/TNk2LxsEAvOLuO52+08zlKssXbc=";
+    };
+    buildInputs = [ pkgs.tbb_2021_11 ];
+    cmakeFlags = attrs.cmakeFlags or [ ] ++ [ "-DBLAKE3_USE_TBB:BOOL=TRUE" ];
```
Right, this is enabled in the pending nixpkgs PR I mentioned.

One difference with the nixpkgs PR is that it is configured to use `tbb` version `tbb_2022_0`. I discuss that in a comment [here](https://github.com/NixOS/nixpkgs/pull/390458#issuecomment-2734805358).

I've instead specified `tbb_2011_11` here because `tbb_2022_0` is not available in the flake `nixpkgs` without updating the refs.

But I would assume I guess that if the packages are updated to where the new `libblake3` version is available, it should be fine to also use the newer `tbb`.

Ericson2314 commented:
@silvanshade https://github.com/NixOS/nixpkgs/pull/390458#issuecomment-2738287520 I left instructions on how to backport things to Nixpkgfs 24.11. Then we can bump the Nixpkgs in the flake, and can do things more simply here.

silvanshade commented:
I've created `24.11` backport PRs for the `libblake3` and `tbb` versions needed here:

https://github.com/NixOS/nixpkgs/pull/391413
https://github.com/NixOS/nixpkgs/pull/391418

Mic92 commented:
Do you have by chance some examples, I could use to check the performance?

Mic92 commented:
I added some patches that I needed to get this build with current nixpkgs unstable. However I think we are currently ending up with two tbb versions somehow? At least I get now crashes on macOS during early inintialization (not just in this pull request but also on master with nixpkgs-unstable).

silvanshade commented:
> Do you have by chance some examples, I could use to check the performance?

I just added a benchmarks section to the original post that gives some details on this.

Performance is an estimate (since it's using the Rust numbers) but practically identical based on my testing locally and original testing upstream for the `libblake3` tbb feature.

Mic92 commented on src/libutil/file-system.cc:
```diff
@@ -243,9 +245,22 @@ std::string readFile(const std::filesystem::path & path)
+    // Memory-map the file for faster processing where possible.
+    if (memory_map) {
+        try {
+            boost::iostreams::mapped_file_source mmap(path);
```
Not sure where is this particular used, but could you run nix-env with and without this change, to make sure we don't add any regressions else where?

silvanshade replied:
From grepping the source and checking references with `clangd` it looks like the only place this particular version of `readFile` is used is in the call from `hashFile`: https://github.com/NixOS/nix/blob/d2ae27e5e53cd023579936408121f2a1885a4699/src/libutil/hash.cc#L353-L358

These are the results I get:

```
hyperfine --warmup 3 'nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null'
Benchmark 1: nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null
  Time (mean ± σ):     363.4 ms ±   1.3 ms    [User: 317.6 ms, System: 42.5 ms]
  Range (min … max):   361.4 ms … 365.1 ms    10 runs
```

```
hyperfine --warmup 3 'nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null'
Benchmark 1: nix-env -f . -qa \* --meta --xml --drv-path --show-trace >/dev/null
  Time (mean ± σ):     363.0 ms ±   1.6 ms    [User: 316.5 ms, System: 43.1 ms]
  Range (min … max):   359.6 ms … 365.2 ms    10 runs
```

It's basically the same. If I run more iterations sometimes one is faster, sometimes the other, but within margin of error.

If there are potential performance regressions from memory-mapping, they would most likely show up on memory-constrained devices like embedded platforms. But it's hard to tell because on average even for memory-constrained devices, most files being read are still probably small enough to fit into memory. Unfortunately I don't have a good system set up to test this on.

If it's a concern, we could add a global configuration option (via `nix.conf`) to allow disabling memory-mapping. We could also change the default for `mermory_map` to `false` and only use it for the BLAKE3 hashes, but for memory-constrained devices (if we are going to account for that) we would probably still want the global configuration option to override that behavior.

Mic92 replied:
I think a smoke test is enough for now.

silvanshade commented:
@Ericson2314 @Mic92 I've updated the `nixpkgs` input with the `libblake3` backport and removed the overrides.

Ericson2314 commented on src/libutil/hash.hh:
```diff
@@ -20,6 +20,9 @@ const int sha1HashSize = 20;
 const int sha256HashSize = 32;
 const int sha512HashSize = 64;
 
+// BLAKE3 data size threshold beyond which parallel hashing with TBB is likely faster.
+const size_t blake3TbbThreshold = 128000;
+
```
Does this need to be exposed in the header?

silvanshade replied:
No, it's not necessary. I can remove it from the header.

It might eventually be useful to have exposed for some other part of the code base but can always revisit that later.

Ericson2314 commented on src/libutil/meson.build:
```diff
@@ -66,13 +66,21 @@ endif
+  method : 'pkg-config',
 )
 deps_private += blake3
+tbb = dependency(
+  'tbb',
+  version: '>= 2021_11',
+  method : 'pkg-config',
+)
+deps_private += tbb
+add_project_arguments(['-DBLAKE3_USE_TBB=1'], language: 'cpp')
```
Is this needed? Shouldn't this dep just be propagated from libblake?

Ericson2314 replied:
https://github.com/BLAKE3-team/BLAKE3/blob/master/c/CMakeLists.txt looking at this, I think some of the logic needs to be updated to also change pkg-config.

Basically I think the rationale in https://github.com/BLAKE3-team/BLAKE3/blob/master/c/CMakeLists.txt#L234C1-L237C65 is exactly right --- yes, you can use a tbb-supporint blake3 build without tbb, but it just runs the risk of linking multiple tbb builds, so better propagate the dep downstream unconditionally.

Ericson2314 replied:
Glad the upstream PR is merged, good luck with the rest!

Mic92 replied:
Btw. we switched in the meantime to unstable.

Mic92 replied:
I think we can rebase this?

silvanshade replied:
We can rebase but I was just waiting for the BLAKE3 changes to settle upstream before making further changes here.

silvanshade commented:
Reviving this PR now that I think most of the `libblake3` build related fixes are in place.

We want to rebase one more time with updated nixpkgs once the very latest 24.11 backported changes (with the MinGW and FreeBSD and i686 fixes) hit the release channel.

Ericson2314 commented on src/libutil/hash.cc:
```diff
@@ -307,11 +307,28 @@ static void start(HashAlgorithm ha, Ctx & ctx)
     else if (ha == HashAlgorithm::SHA512) SHA512_Init(&ctx.sha512);
 }
 
+// BLAKE3 data size threshold beyond which parallel hashing with TBB is likely faster.
+const size_t blake3TbbThreshold = 128000;
+
+// Decide which BLAKE3 update strategy to use based on some heuristics. Currently this just checks the data size but in
+// the future it might also take into consideration available system resources or the presence of a shared-memory
+// capable GPU for a heterogenous compute implementation.
+void blake3_hasher_update_with_heuristics(blake3_hasher * blake3, std::string_view data)
+{
+#ifdef BLAKE3_USE_TBB
+    if (data.size() >= blake3TbbThreshold) {
+        blake3_hasher_update_tbb(blake3, data.data(), data.size());
+    } else
+#endif
+    {
+        blake3_hasher_update(blake3, data.data(), data.size());
+    }
+}
```
I wonder, should this live inside Blake3 itself. Not *instead* of the raw building blocks, but in *addition*. None of the "tuning" that you propose is really Nix-specific, so I think it would be nice if the code lived elsewhere so many projects could share it.

silvanshade replied:
For some additional perspective, this is the relevant section from the [README.md](/BLAKE3-team/BLAKE3/blob/master/c/README.md):

> To get any performance benefit from multithreading, the input buffer needs to be large. As a rule of thumb on x86_64, blake3_hasher_update_tbb is *slower* than blake3_hasher_update for inputs under 128 KiB. That threshold varies quite a lot across different processors, and it's important to benchmark your specific use case.
>
> Hashing large files with this function usually requires [memory-mapping](https://en.wikipedia.org/wiki/Memory-mapped_file), since reading a file into memory in a single-threaded loop takes longer than hashing the resulting buffer. Note that hashing a memory-mapped file with this function produces a "random" pattern of disk reads, which can be slow on spinning disks. Again it's important to benchmark your specific use case.

So the other issue with suggesting some default tuning around this is what values do you select given the above?

Even the Rust version doesn't actually enforce the `128 KiB` threshold, it just mentions it in the documentation. It's up to the user to choose whether to invoke the multi-threaded Rayon update vs the single-threaded update.

Ericson2314 reviewed (commented):
Happy all those TBB build system changes are indeed no longer needed!

silvanshade commented:
I updated the `nixpkgs` input and rebased and now `libblake3` includes the most recent patches (fixes MinGW, FreeBSD, and 32-bit builds).

I believe this resolves all the known issues with failing builds and this PR should be ready for final review/merge.

Ericson2314 reviewed (approved):
OK!
